### ISSUE TITLE
Backfill statuses from remote accounts when first subscribed

### ISSUE THREAD

**Issue by Gargron** [Gargron/mastodon#34]

**Comment by mjankowski** [comment]
I'd like to take a crack at this, but am still trying to wrap my head around the world of services/workers and which things are triggered when.

My guess here is that somewhere in the collection of remote follow services and workers we want to add a background worker task which pulls the feed of the remote account which was just followed, and import the last 5 or 10 or so items.

The other thing I'm struggling to locate is any code which is effectively "go grab this account's atom feed and import some entries" ... it strikes me as possible that this doesn't exist, because all statuses are pushed onto the server and never (yet) pulled?

Any pointers here are appreciated!

**Comment by Gargron** [comment]
@mjankowski The main roadblock to this is actually #1059, because if you went back and imported 5 statuses from an account that last posted 6 months ago, you would get 5 6mo-old posts on the top of the public timeline (and potentially home timelines as well)

So this should probably stay a wontfix until something is done about IDs and their sorting.

**Comment by mjankowski** [comment]
Thanks, I read that thread and looked at more code, and I agree that backfilling is blocked on improving the ordering approach to either get away from ID-sorting, or make ID-sorting reliable for backfilled records. I'll pause on this, and leave a comment over there.

**Comment by deutrino** [comment]
#3307 is about accurate follower counts, but in the comments I discuss some implementation ideas which could theoretically encompass this.

**Comment by jpdeplaix** [comment]
Any updates on this ?

**Comment by Gargron** [comment]
It's now possible to implement this, but it's not clear how many items to fetch (and if you say "all", think about accounts with 100,000 statuses...)

**Comment by jpdeplaix** [comment]
Mmh maybe it can be done by displaying a button where the missing toots are, to fetch the 5 or 10 next toots ? So it would be done manually by users little by little.

**Comment by jpdeplaix** [comment]
@deutrino Generally this seems good, but some toots might be already fetched manually by users and so there would be holes to fill here and there.
For example if there is a profile with no visible toots from one instance and you have access to the urls of some toots and you insert them into the search bar they are going to be fetched and added to the profile of the user in question (from your instance point of view).

**Comment by jpdeplaix** [comment]
@deutrino ok, I was thinking back at what you were suggesting and it seems to be the good thing to do considering that mastodon already serve 20 toots at a time when looking at a profile. So everything can probably be done smoothly on the server side with field saying "I know I'm up-to-date up to this toot/date" and contacting the other instance if we are trying to get toots that are in the "probably not up-to-date" range.

**Comment by jpdeplaix** [comment]
Maybe fixing this would also partly fix https://github.com/tootsuite/mastodon/issues/6137, as looking for a deleted account toots would give hints to the server to remove the account from its local database ?

**Comment by kit-ty-kate** [comment]
Now that #7459 has been merged, I guess it seems simpler to finally implement this. @ThibG ?

**Comment by kit-ty-kate** [comment]
Mmh, thinking again, maybe not. Sorry. They seemed closely related in my head but seems like I'm too tired today :/

**Comment by kit-ty-kate** [comment]
any chance for this issue to get fixed or partially fixed by the next release (2.5) ?

**Comment by ThibG** [comment]
@kit-ty-kate with some logic similar to #7459 we could probably fetch private toots from a remote user, yeah.
Fetching “up to the N last toots” on first follow is easy enough, but:
- it would not be applied to users already followed before the update
- it would not be applied to users not followed
- “up to the last N toots” is pretty arbitrary

The suggestion of having a “gap” that users could click sounds very nice, but it is a *lot* harder to implement. Indeed, the protocol mandates toots to be strictly ordered, but I don't think there is a mechanism to make sure you're not missing toots, nor to request a certain range of items, so efficiently filling a gap seems pretty hard. Add to that that some items may only be displayed to some authenticated users, and things can get quite complicated…

**Comment by trwnh** [comment]
I'm not sure if this deserves a separate issue, but if you follow a locked account and you're the first person to do so on your server, then you can't fetch old followers-only posts. I followed a locked account that had a pinned public post with followers-only replies, where the public post says only "please read:" and the followers-only replies... were not loaded, and thus unable to be read.

I assume this is because the older statuses weren't backfilled, but I'm not sure if there's a different or better way for your server to discover followers-only replies besides backfilling.

**Comment by nightpool** [comment]
ftr, i think you can fetch followers only posts using the search bar now

**Comment by trwnh** [comment]
@nightpool you'd have to know the URL of every single post though, wouldn't you?

**Comment by ThibG** [comment]
Yes, you would need to know the URLs, which you typically don't know when those are *replies* to a toot you can see.
It's a bit of an issue, as those replies are likely to be old posts, and thus not fetched by the “fetch last n toots” proposal.

**Comment by nightpool** [comment]
Ah, my mistake, I thought you meant that the pinned post has a list of
*links* to private posts, not just replies.

**Comment by thebaer** [comment]
Anyone else interested in working on this? Or any advice on where to get started? Would really like to get this feature in (as mentioned in #9525).

**Comment by sikkdays** [comment]
If this issue is too complex to solve or we cannot agree on how to make this work, what about a button or text message telling users to look at the full profile?

It's just my opinion, but discovery is always an issue in social networks. Thus, I find this issue paramount to growing the decentralized and federated world. New people I invite to my Mastodon  instance have no idea that my server has never connected to another and that's why the "toots" section of the profile is empty. Therefore, a simple message in the toots section of a profile that said, "There are no toots here, trying going to the user's page by clicking on their icon above" might be helpful. Something more elegant, I'm just rattling this idea off before I lose it.

**Comment by trwnh** [comment]
You mean like this? blob:https://imgur.com/0499c120-7418-4ef1-80e8-93fb4d25232e

**Comment by ealgase** [comment]
This is probably my biggest gripe with the platform, and I think it would be a roadblock in trying to get my friends to switch over.  It may be hard to implement but it would be VERY useful.

**Comment by GammaPi** [comment]
Downloading all  toots (like relay) or downloading none toots (without relay) are not very good solution as mention above.
So why not implement a "cache on request, delete after some time " feature. I mean user can specify how much toots or how long they want to cache other users' toots. When the user want to check some history toots, the system needs to download and cache it for the user. After sometime, if those toots are not requested again, the server can delete them using a scheduled task.

**Comment by jacotec** [comment]
A fat +1 for the need of this. Tried to find out what's wrong in my instance just because of this ... although I more and more understand Mastodon technically, normal user's won't ... and they won't understand why they see no info when they search for and look at an other user's profile.

The server should grab the needed info on demand from the remote instance hosting that user.

**Comment by kit-ty-kate** [comment]
This is an important issue in my opinion and it shouldn't be closed

**Comment by nightpool** [comment]
(@kit-ty-kate stalebot isn't going to close any issues, the comments / labels were a misconfiguration)

**Comment by kit-ty-kate** [comment]
Well it did for https://github.com/tootsuite/mastodon/pull/8214 and https://github.com/tootsuite/mastodon/pull/9205 at least

**Comment by SherSpock** [comment]
It seems there's a good higher level outline referenced and described by @gammaPi above on what would be wanted for this feature.

One question:
If I am understanding this issue and the last word by @dansup on #1745 properly, is backfilling toots from remote accounts still desired?

> We do not plan to add backfilling on follow because we are disabling outboxes for privacy reasons.

**Comment by ThibG** [comment]
dansup's statement on outboxes is his own, and doesn't affect Mastodon

As far as I know, we plan on continuing supporting outboxes (even though I made a proposal to never include private toots in them, see #13584, which would prevent backfilling private toots when first following a remote user).

As for backfilling statuses, I do think that would be useful (but maybe surprising or unwanted for private toots? this would be consistent with what we do for already-known toots though), but when to do it isn't obvious, and we will never be able to backfill *all* toots, so I'm afraid we'd only create more confusion if we had “batches” of fresh toots, with missing ones between them… that being said it would probably not be much worse than the current behavior…

**Comment by SherSpock** [comment]
Gotcha. Thanks for the info @ThibG! It appears I was conflating statuses and toots. 

Please correct me if I'm mistaken, but based on your reply it isn't clear to me that this is a wanted ticket ready to be worked.

**Comment by ThibG** [comment]
Uh, “statuses” are just what toots are called internally, I used both terms interchangeably, sorry for the confusion.

**Comment by SherSpock** [comment]
Gotcha, my mistake.

**Comment by leo60228** [comment]
I'm not sure what the implementation of this would look like, but would it be possible to seamlessly fetch toots from remote servers without storing them in the local database?

**Comment by mind-overflow** [comment]
hey, any update on this issue?

**Comment by FediVideos** [comment]
Would it at least be possible to backfill the latest 5 posts, for example?

Some of the accounts I recommend on FediFollows are very new, and may have few/no followers, yet they've already posted really nice stuff like their greatest paintings, videos, songs etc, but it won't be seen by anyone because they didn't have followers when they posted them. Even new followers won't see it because there is no backfill.

This is really unfair on new users because even when they make an effort to start by showing their best stuff, it is made invisible due to the lack of backfill.

**Comment by QuestionableTangerine** [comment]
I feel like a lot of suggestions here have been rejected for being imperfect when like, literally anything at all would be an improvement

**Comment by kidwellj** [comment]
> I feel like a lot of suggestions here have been rejected for being imperfect when like, literally anything at all would be an improvement

+1

**Comment by leo60228** [comment]
My understanding is that this mostly hasn't been implemented because it's a lot of effort for something no developers consider a high priority, not because of major implementation questions.

**Comment by ShadowJonathan** [comment]
Then what is high priority? Implementing tiktok-like video sharing? Chasing any other latest media development?

Imo they should focus on their core experience, there's still a bunch of unfulfilled *core* expectancies from other social media apps that just aren't there in mastodon, backfill being one of them, full post migration being another.

**Comment by QuestionableTangerine** [comment]
> My understanding is that this mostly hasn't been implemented because it's a lot of effort for something no developers consider a high priority, not because of major implementation questions.

Fixing this issue would reduce confusion for new users and promote cross instance federation, both of which I would think are high on Mastodon’s priority list! Very unfortunate to hear that is not the case

**Comment by FediVideos** [comment]
Any news on this? It would really help new accounts get going. New people are confused because no one interacts with their first posts, not even followers. And people on smaller instances are seeing completely blank profiles for accounts actually have lots of content.

Even a small amount of backfilling would help a lot with this.

**Comment by mind-overflow** [comment]
I honestly don't understand why we aren't just connecting to the remote mastodon instance and loading content directly from there instead of backfilling other people's data in our own instance. That's how federation is supposed to work. Different servers that each have their own data and can integrate remote data into a uniform platform. No need to download a certain amount of posts or whatever, it feels hacky and still doesn't tackle the issue of having incomplete and incoherent datasets. The web version of Mastodon is way better because you can just browse the profile on its own domain to see all the posts - I don't see why we aren't integrating the remote call when the profile is shown on our domain already.

**Comment by trwnh** [comment]
> I honestly don't understand why we aren't just connecting to the remote mastodon instance and loading content directly from there instead of backfilling other people's data in our own instance. That's how federation is supposed to work. 

Federation works by delivering copies of each status to your followers, generally speaking.

I agree that it is generally "hacky" to backfill statuses, and this "hack" is really only needed because of the paradigm that every interaction must take place inside the (web) app rather than on the remote website. The alternative would be to require cross-domain authentication and authorization, so the remote website knows who you are and so that you can interact natively on the remote website without being kicked back to your (web) app.

The downside, of course, is that unless the wider fediverse adopts this pattern, it will lead to a worse experience. Old private posts would not be visible anymore. Remote interactions would not be possible or easy.

See #7643 for a potential way to do this -- adopting OpenWebAuth or something similar would allow for the needed cross-domain authentication. You'd need to extend it to allow for cross-domain authorization as well, though, and that's trickier since you'd effectively be placing trust in every single website the same way you would have to trust a client application.

**Comment by JPLeBreton** [comment]
> I honestly don't understand why we aren't just connecting to the remote mastodon instance and loading content directly from there instead of backfilling other people's data in our own instance.

Because one of the most important reasons to support post migration is that instances can and do go down forever.

For instances that go down instantly due to hardware failure or the admin pulling the plug and refusing to play nice, there's not much that can be done - hosting is a trust relationship, no way around that.
But for instances that decide to shut down while committing to a reasonable exit process for their users, leaving it up for long enough for all users' posts to propagate to their respective new hosts would be the basic courtesy. This should in fact be considered a basic user freedom - "we will not flush your data unless/until you tell us to, and you can take it anywhere you want with zero data loss". Mastodon cannot currently allow instance admins to make this promise to their users. It should be a cornerstone of federation.

**Comment by FediVideos** [comment]
You can already manually backfill posts by copy-pasting their URLs into your server's search box. Surely there could be an automated system for doing this if someone follows someone? At least for the last five or ten posts?

This is esepcially important for newer or lesser-followed profiles which may look completely blank from some instances.

**Comment by Oblomov** [comment]
Loading at least the most recent toot and the most recent reply, with a button to load more (if any) should be provide a good compromise between UI friendliness and efficiency.

**Comment by nehemiagurl** [comment]
I concur with many of the writers here that automatic backfilling will be an improvement.
However, I want to also raise the issue of backfilling pinned posts, as they often contain important information about the user that new followers may want to know. Also, since Mastodon already puts a hard limit on the number of pinned posts, backfilling them shouldn't cause too much of a server load.

**Comment by FediVideos** [comment]
Had some feedback from a concerned small instance owner, they would like to have some kind of admin control for backfills so they can keep control of resources. Maybe a setting for how many posts to automatically backfill (including zero)?

**Comment by Sobsz** [comment]
> However, I want to also raise the issue of backfilling pinned posts,

The instance I'm on does just that, but it runs on the [glitch-soc](https://github.com/glitch-soc/mastodon) fork, so I don't know if mainline Mastodon does the same.

**Comment by Ember-Ember** [comment]
from what I've experienced, it does not

**Comment by ClearlyClaire** [comment]
Pinned posts should already be backfilled.

**Comment by vanlueckn** [comment]
This would be very nice

**Comment by johnconway** [comment]
I'm going to add my voice to this, but to add that I think it definitely needs to happen _before_ subscription, and should happen on _viewing_. Deciding whether to follow someone back on a different instance at the moment is quite messy!

**Comment by Oblomov** [comment]
An alternative to backfilling for these things could be to be able to view profiles and posts in the other instances without backfilling, but in something like a “nested frame”, with some way to “override” the interaction links so that they are managed within the original browsing context, but I don't think this is possible with XSS protections.

**Comment by nehemiagurl** [comment]
> @johnconway: I think it definitely needs to happen _before_ subscription, and should happen on _viewing_.

That would be nice, but would put a huge and quite unnecessary load on servers to not only fetch user details, but also posts, on every single view.

**Comment by johnconway** [comment]
> > @johnconway: I think it definitely needs to happen _before_ subscription, and should happen on _viewing_.
> 
> That would be nice, but would put a huge and quite unnecessary load on servers to not only fetch user details, but also posts, on every single view.

I don't think load a few posts from a remote server on a profile view would be a huge load – it would be insignificant compared with the federated timeline on most instances surely?

**Comment by trwnh** [comment]
It would be a huge load on the remote server, not on your local server. The
federated timeline just shows you what you have received. Fetching from an
account means the account has to respond to multiple requests, which gets
multiplied by every single server trying to do the backfilling. If a
popular account boosts your tiny account on an underpowered server, it will
be flooded with GET requests for that one post and for its actor. If
backfilling were added naively, then you would be massively increasing this
already massive flood of requests.

**Comment by fkrauthan** [comment]
But wouldn't the server has to give that info + more anyways if I open the profile on the instance the user is on (which most people would do I assume)?

**Comment by johnconway** [comment]
> It would be a huge load on the remote server, not on your local server. The federated timeline just shows you what you have received. 

I was never talking about the load on a local server. The federated timeline is created from remote servers, so the thousands of posts on it have to be requested or delivered from their originating host somehow. In terms of number of posts delivered to instances, filling in a few on profile views has go to be WAY less than contributing to the federated timeline. 

On top of this, if you click the link to view on the original instance, the remote server has to serve the entire profile, including by my count 18 posts. So if only a fraction of people are clicking the link, the load is still higher on the remote.

If this adds 10% to the cost of maintaining Mastodon instances, I'd gladly pay it. Discoverability is SO BAD for new users on small instances.

**Comment by trwnh** [comment]
the federated timeline is created from posts *delivered* to your server.
delivery happens once, and only to followers. what backfilling would do is
add more network requests for statuses that may already have been
delivered, and may be potentially unbounded. i mean, how far back do you
stop? so doing it on profile *views* is absolutely not the way to go

viewing the profile natively on the remote server is still nowhere near the
same. you're only making one request, which gets handled with local data.

i do agree that discoverability is bad, but backfilling is not the only
solution to this. it may still be worthwhile to have backfilling, but in
pretty much every case it would be better to just browse that profile
natively on its home server. the real pain point is not being able to
interact easily, which could be solved by something like cross-domain
authorization. the other pain point is knowing that the profile or posts
exist at all, which could be handled by dedicated discovery features
possibly provided by some kind of hub service.

**Comment by fkrauthan** [comment]
But like I said. Don't we already link: View Timeline on host instance? So people probably going to click on that which is more heavy for the target instance (as they need to serve UI + the data) vs a more lightweight call from instance to instance just transferring data. And we can maybe timebound it (e.g. last X weeks or last X toots only)

**Comment by johnconway** [comment]
I'm sorry, maybe I'm being being really thick here, but I really don't see how loading three, four, or five posts is going to be a massive burden in the target server when visiting that profile on the web would load far more data than that. Loading JSON through the API to the local server should be more efficient all around than loading HTML through the web interface.

I'm not talking about backfilling every post - that would be, I agree, very burdensome.

**Comment by nehemiagurl** [comment]
@fkrauthan @johnconway because the posts already exist on the instance's server and are tagged and organised in a chronological matter, and it is simply a matter of popping open the UI.
If say fedifollows promotes my account on toki.social, if each of the people who saw the post were to visit my profile in toki.social, it is simply a matter of showing them the UI, the posts are already there. If on the other hand we were to backfill ok a view, then for each of the people who clicked through from fedifollows, my instance would have to get all of the view requests, put the recent posts in the outgoing bucket (and change which posts these are every time I post a new one) and transfer them to the other instance. It may be somewhat reasonable to do on a follow since those are more rare (this is on the Devs to say, as they know much better than I do what sort of load comes from such a thing), but on a view would probably be unreasonable.

**Comment by Oblomov** [comment]
The fundamental issue is that when seeing someone's profile on their instance it becomes cumbersome to interact with the posts (like/boost/reply), since you need to get the link, go back to your instance, paste there, and do your interaction. This completely bombs the workflow. Ideally, there should be a way to browse another instance's collection possibly WITHOUT backfilling (like, in a nested context), but still allowing automatic import of the posts users interact with.

**Comment by nemobis** [comment]
> Ideally, there should be a way to browse another instance's collection possibly WITHOUT backfilling (like, in a nested context)

You mean some kind of prefetch? You follow a user, the 10 most recent posts get retrieved and the first 5 displayed. You load the last 5, the next 10 get retrieved. Etc.

It seems v4.0.0 has introduced some kind of infinite scrolling. I don't know whether that makes this prefetching easier or harder.

**Comment by Oblomov** [comment]
> You mean some kind of prefetch?

No, I mean like some nested context. When you open a profile in the webapp, you're opening that profile in the context of your instance, so you only see what your instance knows about it. Ditto for a thread. It would be better if it was possible to open it in “their” context (as it happens when you open it in a separate window), but without leaving the web app (metaphorically, as if it was in an HTML frame; or maybe not just metaphorically?), and then when you interact with content that is presented in the nested context, but not in your instance, it gets automatically imported (as it happens now when you put e.g. the link to an yet-unknown user or post in the search bar), transparently from the user.

**Comment by MikeHuntington** [comment]
I'm seeing a lot of conversation around the web app, and that was fine when Mastodon was less popular. But that the everyday social media browser has learned about it, the MOBILE app is getting hit even harder by this issue. On the mobile app there is no "click through" to the profile. Users are just served an empty profile with no content. It is a very horrible UX on mobile.

I'm sure mastodon admins will tell you, with this influx, mobile sign-ups and users have been the majority.

**Comment by johnconway** [comment]
Yes, this is a UX disaster, for the sake of the cost fetching three or four remote posts occasionally.

**Comment by LeonardoPaccianiMori** [comment]
I second @MikeHuntington and @johnconway 
In the current state this is a UX tragedy that _**will**_ discourage people from using mastodon (if it's not already).
I think it is of _paramount_ importance that this issue is resolved as soon as possible.

**Comment by danog** [comment]
Like https://github.com/mastodon/mastodon-android/issues/359 and https://github.com/mastodon/mastodon-ios/issues/369, this issue details the terrible (lack of?) UX choices that will immediately drive away any potential Mastodon user coming over from the birdsite (or just signing up on their own).

It's really sad that this is the state of federated social networking, where the major federated platform not only has an official mobile app that *doesn't have a federated timeline*, but also both the website and the app simply won't load (even recent) remote user "toots" unless they've been explicitly viewed by pasting the post link in the URL or search bar.

Basic content discoverability and **being able to see other users' public posts** should be a no-brainer for anyone developing a social network.

If mastodon marketed itself as a non-federated, self-hosted social network this would've been just fine, but it's astonishing how broken and hidden away are the so loudly-advertised federation features (that are really just basic features for Mastodon).

I really hope that once this all calms down and users flock back to the birdsite (which isn't going anywhere, especially after its new owner spent 44 bln$ to buy it), the maintainers of this project will sit down and have a little chat about UX and issue prioritization, and come up with a new and improved Mastodon, one that takes itself seriously both as a social network *and* as (the biggest) federated social network.  

And maybe one day, when the EU finally decides to introduce mandatory federation for all social networks, mastodon will be ready to become the biggest and best member of the fediverse.

**Comment by joshbetz** [comment]
Hey all, let's remember that the maintainers are volunteers. I don't speak for the project, but I think it's fair to say that we should try to keep the criticism constructive here.

**Comment by danog** [comment]
As a FOSS maintainer I fully understand that the maintainers are (not exactly unpaid, 26k€/month on Patreon) volunteers, I'm just really frustrated that this is the state of the federated web at such a critical potential tipping point for the web, to the point where even one of the two maintainers of the official mastodon app is developing another [federated alternative to mastodon](https://github.com/grishka/Smithereen) that fixes basic federation features.

I also feel it's correct to call out incomplete and broken features loudly advertised as the main selling point of a software.

I understand that like most FOSS software, this is a project developed by the maintainers for the maintainers, and in this sense I see that mastodon is an excellent self-hosted social network for niche communities, but (probably intentionally at least in the sense of issue prioritization) not a federated social network.

**Comment by danog** [comment]
Not to mention the absence of such a basic feature like the federated timeline in the official app, that was explained in https://github.com/mastodon/mastodon-android/issues/8 (with many more still open issues on the same topic) mentioning app store gatekeeper rules, which makes no sense because the local timeline can still contain boosted illegal content from not-yet-banned instances (and of course you can always directly log into instances that allow illegal content).

Granted, gatekeepers' arbitrary rules can also make no sense, but at least remove the app from the stores instead of removing basic functionality, because yes, a federated timeline is basic functionality, especially now that servers are overloaded and users naturally distribute across smaller servers, finding themselves onto isolated (not interconnected!) islands.

**Comment by babbage** [comment]
As a new Mastodon user, it was confusing to find empty profile pages for most people when I view their profiles to decide whether to follow them or not, after seeing one of their posts boosted onto my timeline.

It was even more surprising to find on the web app that it was possible to click through to a wen hosted profile that had that information, yet it was not displayed when I was viewing their profile logged in via the "main official app" (whether the web app or mobile app).

Reading the thread here has helped me understand how the technical implementation of federation has lead to this situation, and why there are issues to work through to achieve a parity in this.

But as a new user who uses a lot of software but was entirely unfamiliar with Mastodon, let me share my "fresh eyes": it feels nonsensical to view someone's profile in the "Masterdon app" and have it missing the core content, yet see that the missing content is published on the open internet in that user's profile page for non-logged in viewers. From a UX standpoint, this makes no sense, and I warmly encourage the project contributors to recognize their person experience on longstanding instances may be extremely different to new users on newer or even brand new instances.

**Comment by ineffyble** [comment]
> I'm sorry, maybe I'm being being really thick here, but I really don't see how loading three, four, or five posts is going to be a massive burden in the target server when visiting that profile on the web would load far more data than that. Loading JSON through the API to the local server should be more efficient all around than loading HTML through the web interface.

If it's being delivered in the same manner as regular posts, then that involves negotiation between instances and establishing trust using HTTP Signatures. Meaning multiple requests and database activity on both sides.

**Comment by ClearlyClaire** [comment]
In addition to the cost of fetching the latest N posts themselves (and properly handling “holes”), there is another limit with Mastodon's current design overall: it has to resolve any reference (mentions, parent post, …) for any object it fetches, because it has no concept of unresolved references. This means increasing the amount of content it backfills will also increase the amount of content it will have to recursively backfill, making the limits in #22025 much easier to hit (hitting those limits means some references will be left unresolved).

**Comment by cotsuka** [comment]
Something to consider is implementing this first as a feature only for single-user instances. Those are the instances that, to me, would most benefit from this. I've recently stood up my own single-user instance and realized it's basically useless since I need to constantly view all of my follows' profiles on their home instance to see anything.

**Comment by nehemiagurl** [comment]
> Something to consider is implementing this first as a feature only for single-user instances.

The only problem is you can't really differentiate between a single-user instance and a multi-user instance. A single-user instance can have bots run by that user, or past users who stopped being active and now there's just one (the latter actually happened to me).

**Comment by PinballsWizard** [comment]
> Something to consider is implementing this first as a feature only for single-user instances. Those are the instances that, to me, would most benefit from this. I've recently stood up my own single-user instance and realized it's basically useless since I need to constantly view all of my follows' profiles on their home instance to see anything.

I understand there are technical difficulties with limiting things to a single-user instance, but as I run one as well I run into this same problem and it makes mastodon less appealing to suggest for people to run themselves, and I would greatly appreciate being able to backfill statuses.

**Comment by jakewinter1** [comment]
It is beyond embarrassing @Gargron that this is unresolved at this stage. It makes running a small server near impossible

**Comment by kostajh** [comment]
> It is beyond embarrassing @Gargron that this is unresolved at this stage. It makes running a small server near impossible

As frustrating as the issue is, the maintainers don't owe you anything, so please try to be constructive.

**Comment by FediVideos** [comment]
> > I'm sorry, maybe I'm being being really thick here, but I really don't see how loading three, four, or five posts is going to be a massive burden in the target server when visiting that profile on the web would load far more data than that. Loading JSON through the API to the local server should be more efficient all around than loading HTML through the web interface.
> 
> If it's being delivered in the same manner as regular posts, then that involves negotiation between instances and establishing trust using HTTP Signatures. Meaning multiple requests and database activity on both sides.

Pinned toots are already backfilled, so that is possible.

Is it not possible to extend this to at least the latest five posts as well, for example?

**Comment by BBaoVanC** [comment]
Referring to https://github.com/mastodon/mastodon/issues/34#issuecomment-1338999322:

> there is another limit with Mastodon's current design overall: it has to resolve any reference (mentions, parent post, …) for any object it fetches, because it has no concept of unresolved references

Is making Mastodon able to hold unresolved references a feasible way towards fixing this backfilling issue?

**Comment by smiba** [comment]
> Pinned toots are already backfilled, so that is possible.
> 
> Is it not possible to extend this to at least the latest five posts as well, for example?


Considering we already have a collection service for pinned posts, surely we could make a similar service (or modify it) to also try to retrieve a bunch of recent posts? 

https://github.com/mastodon/mastodon/blob/af578e8ce0aabdbe9c0cd3d72d6fa2cc30b7fc66/app/services/activitypub/fetch_featured_collection_service.rb

This seems to be a recurring issue for new users, but even as an experience user I think constantly having to open a new browser tab just to quickly glance over someones profile is tedious

**Comment by FediVideos** [comment]
There's been a discussion on Masto about backfilling replies on threads, as people can't see all the replies. It's not quite the same as backfilling after subscription, is this worth adding as a separate issue? (I don't want to add issues unnecessarily, which is why I'm asking)

**Comment by GnomedDev** [comment]
What would be the problem with:

For the first N requests, fetching the data required for the user's interaction (viewing replies, viewing past user posts) from the remote instance.
Once N requests have been made, properly backfill the data and add a TTL to prevent runaway resource usage?

I assume fetching just the data required would be hard? but I'm not familiar with ActivityPub

**Comment by Shanesan** [comment]
> There's been a discussion on Masto about backfilling replies on threads, as people can't see all the replies in certain circumstances. It's not quite the same as backfilling after subscription, is this worth adding as a separate issue? (I don't want to add issues unnecessarily, which is why I'm asking)

I feel these issues are one and the same, really. One is gathering data when browsing through someone you're _following_, one is gathering data when browsing through a thread you are _trying to follow_. I believe ActivityPub would be using similar backend tools to do the same thing.

**Comment by Oblomov** [comment]
Agreed. These are two faces of the same coin: Mastodon caching remote content locally, and only showing the cached content without providing any means to pull more content in the cache without explicitly requesting posts one by one (through the search box). Both could be solved by either

* manually triggering a request for more context (e.g. refresh of remote user account + some backfilling of recent posts; or cache full threads);
* display remote content without caching (until preferrable).

I've seen it discussed on Mastodon that doing the latter would “violate” the spirit of Mastodon's approach to the protocol, but  I disagree: if I have to “Open remote page” on a separate browser tab, the remote server is getting the same amount of “stress”, _and_ my experience is worsened because if I want to interact with any of the additional content I have to do the copy-link-before-interacting dance.

This could be solved the clients (including the PWA and the “official” apps) providing an option to manually load content without going through the server cache until/unless the user interacts with the additional content. Server support for this would be a plus.

**Comment by sysadminpower2019** [comment]
can we just find a solution and implement it. We are losing a lot of non techinical / new users over this topic. People are confused and overwhelmed. 
I don't think it matters in the end what solution we implement as long as we implement something. It may make some people upset but what is the point of maintaining purity when you don't have users. 
My personal opinion simply allow the instance owners to decide how much to backfill / allow users to select a backfill option for a new profile.

**Comment by nellie-m** [comment]
> can we just find a solution and implement it. We are losing a lot of non techinical / new users over this topic. People are confused and overwhelmed. I don't think it matters in the end what solution we implement as long as we implement something. It may make some people upset but what is the point of maintaining purity when you don't have users. My personal opinion simply allow the instance owners to decide how much to backfill / allow users to select a backfill option for a new profile.

I can confirm. I'm a pretty non-technical person, and set up my own instance with a hosting service. I was prepared to see people's empty profile pages. I was prepared to see an empty local timeline. With my following list migrated, my home timeline looks fine - until I open posts to see more comments, answers or the original first post they refer to. 

I was NOT prepared to have to have a browser window open with the thread I want to read, and have to fetch Every. Single.Comment. into my timeline via copying and searching on my account. Even of accounts I follow. From being a happy member of home.social (shutting down soon) I have moved to being an unhappy user sitting on a desert island of one, with only faint echos of the conversations out there coming through the fog.
Most unhappy, to be honest.

I am insanely in love with the idea that people, families, small groups could, in theory, have their own little homes on the fediverse, and with Mastodon in particular. But in practice, this isn't an option without backfilling the threads folks want to read and comment on. You're just not part of the party without it.

Having read some of this thread here, and from what I understand of it, one of the concerns about backfilling, and one of the reasons why it still isn't implemented, is that it would put undue strain on servers. 

Now guess what recommendations I get for my problem: "use relays".

**Comment by PinballsWizard** [comment]
> I can confirm. I'm a pretty non-technical person, and set up my own instance with a hosting service. I was prepared to see people's empty profile pages. I was prepared to see an empty local timeline. With my following list migrated, my home timeline looks fine - until I open posts to see more comments, answers or the original first post they refer to.
> 
> I was NOT prepared to have to have a browser window open with the thread I want to read, and have to fetch Every. Single.Comment. into my timeline via copying and searching on my account. Even of accounts I follow. From being a happy member of home.social (shutting down soon) I have moved to being an unhappy user sitting on a desert island of one, with only faint echos of the conversations out there coming through the fog. Most unhappy, to be honest.

As another single user instance user, one suggestion that has not come up in this issue yet that you may find useful as I have, would be to run https://github.com/nanos/FediFetcher

**Comment by timnolte** [comment]
I have been using the FediFetcher job and have not noticed anything super major. I suspect I might be missing some parts of conversations occasionally as I will see what looks like a response to someone but no parent message to it. This seems like a rarity though. One other thing that can also be done when you have a single user instance is to use a relay to feed into your instance. Setting up various relays has done wonders for my instance and it has been pretty great.

**Comment by nehemiagurl** [comment]
I agree with @timnolte that relays is the better (and the intended) way to go.

**Comment by smiba** [comment]
> I suspect I might be missing some parts of conversations occasionally as I will see what looks like a response to someone but no parent message to it. This seems like a rarity though.

@timnolte This shouldn't be possible, mastodon always revolves the entire message chain. Unless someone deleted their comment, you should be able to see what a reply was to.

**Comment by ryliejamesthomas** [comment]
> > I suspect I might be missing some parts of conversations occasionally as I will see what looks like a response to someone but no parent message to it. This seems like a rarity though.
> 
> @timnolte This shouldn't be possible, mastodon always revolves the entire message chain. Unless someone deleted their comment, you should be able to see what a reply was to.

He's probably blocked.

**Comment by nehemiagurl** [comment]
> Unless someone deleted their comment, you should be able to see what a reply was to.
Not necessarily. If someone is replying with followers-only privacy, only their followers and whomever is mentioned in the reply will be able to see it, and if for any reason OP's instance refuses fetching (e.g. a non-Masto instance running in limited federation mode, defederation etc) you won't see the replies either. And of course there's just regular user-level muting and blocking.
In fact, if you see a only parts of a reply chain but not others, I suspect either of these scenarios is more likely, since Mastodon does try and fetch the reply chain when you open a reply.

**Comment by timnolte** [comment]
To expand on my earlier comment about possibly missing some parts of a conversation here is what I've observed. I will see someone mention someone(and usually, if not always, someone else that I follow with what looks like a reply to something they said. When I try to look at the thread I don't see an original/starting post. It's very possible that the post I'm seeing is not actually a reply, or as others here have mentioned that perhaps that starting post was deleted.

Again, with the setup I have now these seemingly missing conversation parts are an extreme rarity.

**Comment by nehemiagurl** [comment]
@timnolte as several people said before, this likely is not due to missing functionality in Mastodon (because Mastodon does pull all available posts a reply is replying to) but it's simply not visible to you for any other myriad of reasons (post visibility, deleted etc).

**Comment by timnolte** [comment]
@nehemiagurl yeah, that is essentially what I was getting at. Which is why I noted it's an extreme rarity.

**Comment by FediVideos** [comment]
Is there any progress on backfilling? Just had to deal with someone who thinks all other accounts are inactive because they can't see any activity on those accounts from their smaller server. This happens a lot, and it seems to put people off using Mastodon.

If it's a question of resources, why not let the instance admin decide how much is backfilled? They could set it to be completely off, or backfill a specific number of posts, or backfill everything.

**Comment by melroyvandenberg** [comment]
People are leaving Mastodon because of this missing backfilling feature. I'm not joking devs.

**Comment by saschanaz** [comment]
Mastodon already have AP outbox implementation: https://mastodon.social/users/Gargron/outbox?page=true

I guess technically the only thing that should be implemented is to read it from the remote servers.

**Comment by saschanaz** [comment]
Lemmy supports fetching outbox: https://github.com/LemmyNet/lemmy/blob/f5209fffc1de527db7ea007d463c158b36fda515/crates/apub/src/objects/community.rs#L141-L143

It may be worth looking at their implementation and taking some knowledge.

**Comment by sysadminpower2019** [comment]
i think not just backfilling. Realistically the best idea would be to reach
out to average users and particularly twitter users to see what confuses
them and then adapt to their needs.
We need to be practical, what good is a social network without any people
on it. If we want mass adoption then we need to hide the complexity for the
average user and enable advanced user to customize the experience.
The issue is as always the core devs wanting to maintain their own purity
to the ideology.

**Comment by dannymate** [comment]
First day with fediverse apps. Previous experience with MicroBlogging basically zilch. I used both Mastodon and CalcKey. Here was my experience with this (I do stuff wrong but that's the point. This is what people will do.):

1. I navigated to a remote user using the address bar. This was obviously unintuitive, I don't think anyone wants be messing around up there.
2. When the profile loaded I wanted to view their posts. So I go to scroll down and only some seemingly random posts from days/months/years ago are there. I try to scroll further expecting there to be some kind of infinite load, there is not. So I actually think that's all they've posted. This actually occurred a couple times. (Yes I ignored the pinned post text).-

-> It's worth noting here I think most people would assume Mastodon/Fediverse is pretty empty and full of old accounts. Especially on a small instance like I was on.

3. I stumbled across an external guide that explained that only pinned & explicitly viewed posts show up on remote profiles. So I immediately realised those old posts I was seeing were pinned.
4. I navigate back to a remote profile. I then follow expecting the profiles posts to magically appear. They did not. 
5. (This was CalcKey I'm unsure if Mastodon has an equivalent) I click a button to navigate to the profile on the remote instance and I can see their posts. 
6. I then read that you only get the new posts of people you follow. I think to myself "fediverse things".
The end.

I have some other experiences if it's useful but this is the relevant one to this discussion.

**Comment by saschanaz** [comment]
I think advocating is enough as everyone really should know this is a problem.

Per my understanding Mastodon supports outbox paging (which is required to fill the gap described in https://github.com/mastodon/mastodon/issues/34#issuecomment-392873406), but AFAICT this is not a standard and implementations use their own parameter names for this. ([Mastodon uses `min/max_id`](https://github.com/mastodon/mastodon/blob/4fe2d7cb59f4622ff8af2f048b883f413e87c68e/app/controllers/activitypub/outboxes_controller.rb#L55-L61) while [Misskey uses `since/until_id`](https://github.com/misskey-dev/misskey/blob/a1327fa9e1329f2fb00d70b1e2332cea015bfdee/packages/backend/src/server/ActivityPubServerService.ts#L325).) Some implementations may not support such parameters at all, as seemingly Lemmy doesn't support either of such options. In other words, we don't have infra for this and we need to solve it first.

**Comment by SushiKishi** [comment]
Vaguely related to this issue: could the message at the bottom of a feed be changed?

When I just follow someone, I get 1 (one) pinned message from a year ago. Their toot form six minutes ago does not appear. As someone who has used non-federated social media for a long time now, that certainly doesn't make sense to me. The only message I get that's helpful is this at the bottom of the user's feed I just followed:

> "Older posts from other servers are not displayed.
> [Browse more on the original profile](https://vt.social/@VTSocial)"

OK, so..."older posts from other servers are not displayed." THis pinned toot from a year ago is here. Is that toot from six minutes again somehow deemed older, because it would be "under" the pinned tweet? That doesn't make sense, by that logic only the pinned toot would ever be displayed. That can't be the intent. There must be something wrong with my instance of Mastodon. I must have screwed something up.

What other conclusion was I supposed to come to form that sequence of events?  Why not have the message say:

> "Older and non-pinned posts from before following users from other servers are not displayed.
> [Browse more on the original profile](https://vt.social/@VTSocial)"

or

> "Older and non-pinned posts from before discovering a user on this server are not displayed.
> [Browse more on the original profile](https://vt.social/@VTSocial)"

Then I'd immediately see the problem even if I had no idea what a federalized whatzit website is supposed to be or what to expect. "Oh, it grabbed the pinned post because that's presumably important, but it doesn't get anything else. That's maybe not what I wanted or expected, but at least I know it's not something I did wrong, that's just how it works. Weird!"

**Comment by saschanaz** [comment]
>Vaguely related to this issue: could the message at the bottom of a feed be changed?

A new separate issue for that problem could be helpful.

**Comment by dannymate** [comment]
> I think advocating is enough as everyone really should know this is a problem.
> 
> Per my understanding Mastodon supports outbox paging (which is required to fill the gap described in [#34 (comment)](https://github.com/mastodon/mastodon/issues/34#issuecomment-392873406)), but AFAICT this is not a standard and implementations use their own parameter names for this. ([Mastodon uses `min/max_id`](https://github.com/mastodon/mastodon/blob/4fe2d7cb59f4622ff8af2f048b883f413e87c68e/app/controllers/activitypub/outboxes_controller.rb#L55-L61) while [Misskey uses `since/until_id`](https://github.com/misskey-dev/misskey/blob/a1327fa9e1329f2fb00d70b1e2332cea015bfdee/packages/backend/src/server/ActivityPubServerService.ts#L325).) Some implementations may not support such parameters at all, as seemingly Lemmy doesn't support either of such options. In other words, we don't have infra for this and we need to solve it first.

I was doing some research.which led me to the Fediverse Enhancement Proposal  (FEP)process. Which is currently relatively small but tries to create proposals for interoperability problems within the Fediverse.

When I [mentioned the problem](https://codeberg.org/fediverse/fep/issues/119#issuecomment-960264) you described I was pointed to [this FEP](https://codeberg.org/fediverse/fep/src/branch/main/fep/5bf0/fep-5bf0.md). I'm not knowledgable enough yet to understand if it suits this use case. So I thought I'd bring it directly to you.

**Comment by relgames** [comment]
Found this issue while searching for "Older posts from other servers are not displayed". It is really annoying/confusing, especially in the app where even that message is not displayed, it just shows "0 posts" so it made me thing the app or the server I'm on was broken / buggy.

Seeing this bug open since 2016, I guess it won't be fixed. Too sad, I liked the federation idea, reminds me of old FIDO times. But with people joining and trying to connect with others and seeing empty profiles in the app, it will make them leave.

**Comment by SkaveRat** [comment]
Did anyone ever implement a proof of concept for this that can be tested out?

It's a huge problem for new users who would love to see posts of their friends, but just aren't able to. Making them turn away from mastodon or fediverse. Happened to multiple people I personally know already

**Comment by smiba** [comment]
I do want to state that I also agree that this is a massive issue, scaring people away from the fediverse / mastodon. I personally know multiple people who got confused by this and didn't fully trust the platform anymore after

**Comment by limdingwen** [comment]
A few practical solutions here:

* Has any Mastodon server or fork already implemented this feature? If so, can anyone contribute a link to such patches?
* Anyone here have the time to help implement the feature? Even if it is not merged, server admins can apply it by themselves. Furthermore, it increases the chances of a merge.
    * I do not have the time.
    * Any technical blockers?
* If nobody here has time to do so, may I suggest, for those who want this feature, to open a bounty on this issue? It's like crowdfunding a "feature commission".
    * I know BountySource. Any other alternative platforms that support GitHub?
    * I do see now that Mastodon gGmbH already has funds for this, but my point is even if they're unwilling, it's possible to "commission" a non-affiliated developer to do this.

**Comment by NepNep21** [comment]
I don't recommend posting bounties on BountySource, due to bountysource/core#1586.

**Comment by FediVideos** [comment]
Any progress at all? As far as I can tell, this has been the most requested feature for the last few years?

We already know it's possible via pinned posts, could this not be extended to the latest five non-pinned posts? The burden on other servers would be identical to the current pinned backfilling.

The lack of backfilling is punishing new users by making it much harder for them to be visible, and they are giving up before building up a network of followers that would make them visible. The lack of backfilling is hiding people's first steps onto Mastodon, which is not a good first experience.

**Comment by rkingett** [comment]
What about backfilling 50 toots not already fetched and then have the user control the rest of the fetching?

**Comment by Shanesan** [comment]
> What about backfilling 50 toots not already fetched and then have the user control the rest of the fetching?

The number of toots won’t matter since this will surely be up to the instance owner; many hundreds or thousands of external requests/second from federated instances is likely the issue here which is likely the bigger hurdle for implementation: “Rush hour” requests could literally take down an instance feeding other instances, much like Los Angeles highways at any time of day. 

Making people push the triple dot and click “show original” for replies, or “Open in Browser” or whatever for Profiles, slows engagement to a crawl which makes the service useless but stable. The system needs a limiter developed to make sure servers with large user bases can handle the stress of all these other instances requesting information on-demand.

**Comment by SkaveRat** [comment]
Throtteling resource requests is not something new.

The hosting instance should be able to set a maximum of posts that another instance get fetch per timeframe.
Let's say 100 per hour per user. And if the maximum is exceeded, a client should either gracefully back off or check a response(-header) when to retry again.


So `instance1.example.com` hosts `@famoususer@instance1.example.com`. Someone from `instance2.example.org` follows that user, and as no one else on that instance follows that user, `instance2.example.org` starts fetching 100 posts of that user.

Now there are two ways: Either automatically fetch the whole backlog of posts, or do so on demand if someone scrolls through the timeline of `@famoususer`.


The biggest problem I have with the "click here to see all posts" workaround right now is, that it's basically impossible to see older posts by "locked" accounts (followers-only, follower-approval active)

**Comment by Corgi** [comment]
> The biggest problem I have with the "click here to see all posts" workaround right now is, that it's basically impossible to see older posts by "locked" accounts (followers-only, follower-approval active)

I find myself here for that very issue---a good friend of mine prefers to keep their accounts locked from the public, and while he's added me as a follower I'm still unable to see his old posts since we're on different instances. It's okay for this specific case because he hasn't posted much yet, but I'm sure I'll feel the bite of this as I become more active on Masto and follow more people.

**Comment by JCViscera** [comment]
I made an account on Github just to add my support for this addition and to comment on it. I'm a relatively new user on Mastodon, and I must say, it's incredibly confusing, annoying, and discouraging to necessarily have to go and find another account on a separate instance/domain just to see what posts they've made before. I understand that the point is to not have to host a lot of data, but am I going to necessarily have to refigure out where the public posts of another user are every time I want to see them?

I should also add that this is doubly annoying because some instances don't allow you to do certain things (like view certain posts) unless you have an account on there. What if I want to favorite a post that isn't on my instance and also is not loaded on my instance? I don't want to have to create multiple accounts on Mastodon just because posts that I want to go through someone's backlog to favorite aren't available on the particular instance my account is on. I specifically favorite posts to look back on and find later, so it really breaks a big way that I currently use social media.

If backfilling was ever implemented, I feel there could be a couple options:
- When a user on one instance follows someone located on another instance, backfill a set number of posts of that followed profile to the current instance. The number of posts to be backfilled could be a number set by the owner of the instance (25, 50, 100, etc), and if the owner doesn't want to allow backfilling, then they can turn off the option altogether/set the number to 0. Could also maybe include an option for amount of data retrieved per profile (1MB, 10MB, etc) if the profile includes a lot of images/video in their posts.
- Backfills of public accounts are able to be seen by anyone on the current instance. If an account is Locked, they would only be "Unlocked" to the person who has had their follow accepted and only on the original page. This would be for privacy purposes, since if locked/private accounts were downloaded to the instance's owner's server, this would be an obvious security issue. A flag to allow or disallow this would be able to be set by the owner of each instance. The user would also be notified of this when their follow is accepted.
- In the same vein, backfills of public posts are able to be seen by anyone on the current instance. If an account has private posts on it, they would only be viewable to the person who has had their follow accepted and only on the original page. This would again be for privacy purposes, etc. A flag to allow or disallow this would also be able to be set by the owner of each instance. The user would also be notified of this when they follow someone with private posts for the first time.
- If some user spams or follows a lot of accounts at once (either intentionally cuz they're going on a binge run or disingenuously), only a set number of accounts will be backfilled per day. This number per day (or potentially per account) could be set by the owner.
- If a user did follow a bunch of accounts disingenuously, the owner could either choose to not have accounts followed by them be backfilled or ban the user from the site altogether. Could be a flag or something to set.

I really want to give this place a chance, because I'm really missing a social media that has a kind of community atmosphere to it, but this issue in specific really discourages me from staying, as I do go back through people's profiles often when first following them. Please at least consider the option. Thank you!

**Comment by tesaguri** [comment]
> * When a user on one instance follows someone located on another instance, backfill a set number of toots of that followed profile to the current instance. The number of toots to be backfilled could be a number set by the owner of the instance (25, 50, 100, etc), and if the owner doesn't want to allow backfilling, then they can turn off the option altogether/set the number to 0. Could also maybe include an option for amount of data retrieved per profile (1MB, 10MB, etc) if the profile includes a lot of images/video in their toots.

Couldn't we just use the existing rate limit machinery instead of introducing a new system?

> * Backfills of public accounts are able to be seen by anyone on the current instance. If an account is Locked, they would only be "Unlocked" to the person who has had their follow accepted and only on the original page. This would be for privacy purposes, since if locked/private accounts' toots were downloaded to the instance owner's server, this would be an obvious security issue. A flag to allow or disallow this would be able to be set by the owner of each instance. The user would also be notified of this when their follow is accepted.
> * In the same vein, backfills of public toots are able to be seen by anyone on the current instance. If an account has private toots on it, they would only be viewable to the person who has had their follow accepted and only on the original page. This would again be for privacy purposes, etc. A flag to allow or disallow this would also be able to be set by the owner of each instance. The user would also be notified of this when they follow someone with private toots for the first time.

The concern about visibility is not unique to backfilling. It applies to manual fetch of posts by putting URL in the search field, for example, and so I guess the necessary precautions are already implemented.

> * If some user spams or follows a lot of accounts at once (either intentionally cuz they're going on a binge run or disingenuously), only a set number of accounts will be backfilled per day. This number per day (or potentially per account) could be set by the owner.
> * If a user did follow a bunch of accounts disingenuously, the owner could either choose to not have accounts followed by them be backfilled or ban the user from the site altogether. Could be a flag or something to set.

Again, doesn't rate limiting of following suffice? Perhaps, we might want to reconsider the default limit value based on the increased cost, though (given that we're actually going to backfill when following an account instead of e.g. when the user explicitly requested to backfill (<https://github.com/mastodon/mastodon/issues/34#issuecomment-368341243>) or when scrolling a user timeline (<https://github.com/mastodon/mastodon/issues/34#issuecomment-368345847>) as suggested before).

**Comment by tesaguri** [comment]
Well, the issue has gotten really long and it became more and more hard to grasp the points and blockers.

We need a summary of existing discussions, I suppose? Maybe I can make a write-up if I find time.

**Comment by trwnh** [comment]
closest thing to that: https://github.com/mastodon/mastodon/issues/34#issuecomment-1229115392

points:
- you can either fetch from the outbox on-demand (backfilling), or you can browse natively on the remote site with cross-domain authentication (OpenWebAuth, IndieAuth, etc)
- fetching on-demand requires a remote outbox to be served consistently with respect to paging, and has the downside of not knowing where any gaps may be: https://socialhub.activitypub.rocks/t/stricter-specifications-for-pagination-of-collections-and-orderedcollections/2633
- cross-domain authentication requires both sites/servers to support the same authentication scheme. see #7643 or #24066 for more discussion of the two leading options

the current plan i think is to have a button on profiles that lets you fetch more statuses on-demand, rather than having it be automatically done on follow, but there's issues with doing this consistently as mentioned above. the only truly consistent way to do it is to fetch the entire outbox every single time. the next best thing is to start at the beginning and keep loading new pages until you start seeing stuff that you haven't seen before.

**Comment by jakewinter1** [comment]
“Connecting to the remote mastodon instance and loading content directly”

This exactly what Mammoth does and it’s a really clean and seamless experience.

**Comment by tesaguri** [comment]
> “Connecting to the remote mastodon instance and loading content directly”
> 
> This exactly what Mammoth does and it’s a really clean and seamless experience.

Ah, a client feature to fetch the outbox is an important alternative indeed. I'm not sure that the original comment actually meant that, but maybe client-side fetch and server-side proxying (and possibly cross-domain auth) in combination should sufficiently represent "Connecting to the remote mastodon instance and loading content directly," I suppose.

I've updated the list to replace the item with that one.

**Comment by tesaguri** [comment]
I know the issue has already had enough discussion about the motivation for the feature, but I'd like to add a new one that has not been mentioned yet, concerning privacy.

While you can open the profile page on the original server to browse a remote user timeline, directly connecting to the original server means that your fingerprinting information like IP address is exposed to the server. And theoretically speaking, if you interact with the remote account using your account shortly afterward, the administrator of the server might be able to link the fingerprint with your account.

That might sound irrationally nervous, but with the upcoming participation of closed-sourced corporate implementations like [Tumblr][tumbler-ap-support-intention], [Post News][post-ap-support-intention] and [Threads][threads-announcement] (of the famous Meta Platforms Inc., you know) in the Fediverse, I expect the problem of tracking will become more serious.

And in connecton with this motivation, I'm concerned about the use of authorized fetches in backfill. While using the user's private key to fetch outboxes enables fetch of non-public posts, that reveals the user's identity to the remote server. If the backfilling is to occur when the user followed an account, that's fine since to follow an account essentially involves revelation of identity to the followed account. But if the backfilling is to automatically occur when just scrolling through a user timeline, it is arguably not the user's intention to tell the remote server (possibly operated by the owner of the remote account) that it is the user who is scrolling through the remote account's user timeline right then.

So I think that the server actor's key should always be used to backfill timelines at least if the user who triggered it is not following the target account, and even if the user is following the account, there should be a setting to opt in/out of the use of the user's key, or at least some sort of warning message in the UI.

Also, we can use the key of a randomely selected follower of the account (similar to #7459) to "anonymize" the requester, but that's only a mitigation and doesn't help at all when you are the only follower of the account on the server (or if the account's followers are very few), and I don't expect it would work properly for remote servers with a novel visibility support like "group-only".

[tumbler-ap-support-intention]: <https://twitter.com/photomatt/status/1594577983028740096>
[post-ap-support-intention]: <https://post.news/@/user/2JADKR9kk2IcqtC0Aa6EimpRRjs>
[threads-announcement]: <https://about.fb.com/news/2023/07/introducing-threads-new-app-text-sharing/>

**Comment by Williamalong23** [comment]
> Mastodon already has AP outbox implementation: https://mastodon.social/users/Gargron/outbox?page=true
> 
> I guess technically the only thing that should be implemented is to read it from the remote servers.

✅ Verified in all 🇱🇷 services, Recover Bitcoin, iCloud bypass, Bank loader, game hacks ,Wi-Fi recovery ,Instagram , Facebook ,Snapchat , Twitter ,Gmail, cashapp, PayPal, and other hacking job👌Contact Him 22 on Instagram for quick recovery now https://instagram.com/22cracker?igshid=YWYwM2I1ZDdmOQ==

**Comment by Williamalong23** [comment]
> Lemmy supports fetching outbox: https://github.com/LemmyNet/lemmy/blob/f5209fffc1de527db7ea007d463c158b36fda515/crates/apub/src/objects/community.rs#L141-L143
> 
> It may be worth looking at their implementation and taking some knowledge.

✅ Verified in all 🇱🇷 services, Recover Bitcoin, iCloud bypass, Bank loader, game hacks ,Wi-Fi recovery ,Instagram , Facebook ,Snapchat , Twitter ,Gmail, cashapp, PayPal, and other hacking job👌Contact Him 22 on Instagram for quick recovery now https://instagram.com/22cracker?igshid=YWYwM2I1ZDdmOQ==

**Comment by brendanjones** [comment]
What about follows/followers? Wouldn't the moment that statuses are backfilled be a good time to also fetch the follows/followers of the account? 

Assuming of course it's not already done at the moment you first view the profile, as requested by https://github.com/mastodon/mastodon/issues/19880 and https://github.com/mastodon/mastodon/issues/20533. 

As suggested in those issues, I think it better to always be able to see the full follows/followers list of an account, but if for some technical reason that can't or shouldn't be done, then grabbing the list when you follow an account seems like a good second-best solution.

**Comment by smiba** [comment]
The struggle with this is that Mastodon doesn't like unresolved items (which is sensible), meaning that by resolving this list, it will also need to load and store information for all these profiles (like bio, profile picture, profile name). And as there is no other connection to these accounts, they will grow stale and show incorrect information (name doesn't update, profile picture doesn't update etc.) 

There is no real way to solve this, other then store and exchange massive amounts of information that are very unlikely to be ever seen by any users. (And this times 100,000, for every instance of mastodon that exists) 

This is simply due to the federated nature of mastodon, and the zero-trust approach.

**Comment by zakwilson** [comment]
So it seems to me there's mostly consensus that this should be done, some debate over the details, and a lack of resources to actually work on it.

I know Rails. I like Mastodon. I want this to happen. What behavior would actually be accepted if implemented in an otherwise-satisfactory PR? My thinking is that a page of the ActivityPub outbox should get loaded, and if the user scrolls to the bottom, the next page should get loaded.

**Comment by brendanjones** [comment]
Pinging @Gargron , there’s^^ someone offering to do the work. Are you able to define requirements/acceptance criteria?

**Comment by smiba** [comment]
It'd an open source repo, I assume it will get accepted as PR but even if it doesn't, if it works fine I'd be more then happy to adopt it for my instance.

@zakwilson If you're up for making this still, I'm confident you'll make a lot of users happy!

**Comment by Sythelux** [comment]
At this point I can only recommend @zakwilson make a ready PR at the glitch-soc repo (https://glitch-soc.github.io/docs/ ) and talk with them on how to get an upstream PR to here. This discussion is going on for so long that there won't be any more movement back an forth.

**Comment by FediVideos** [comment]
> At this point I can only recommend @zakwilson make a ready PR at the glitch-soc repo (https://glitch-soc.github.io/docs/ ) and talk with them on how to get an upstream PR to here. This discussion is going on for so long that there won't be any more movement back an forth.

If Glitch introduced some kind of backfilling, think a lot of instances would switch over to it.

**Comment by ShadowJonathan** [comment]
Changes from glitch-soc often upstream to mastodon proper, in many ways its a `develop` branch of mastodon where features get beta-tested in production.

**Comment by Kleisli** [comment]
> I know Rails. I like Mastodon. I want this to happen. What behavior would actually be accepted if implemented in an otherwise-satisfactory PR? My thinking is that a page of the ActivityPub outbox should get loaded, and if the user scrolls to the bottom, the next page should get loaded.

@zakwilson did you ever start working on this?
I think the main point to improve UX is to see the latest posts **before** following an account. Profiles look abandoned and it's not even surprising because many actually are abandoned, but this is a major problem for accounts that aren't.

**Comment by trisweb** [comment]
@Kleisli @zakwilson I agree, visiting empty profiles on the home server is the jarring experience, regardless of whether you newly followed or not.

I think there are a few levels to a potential solution:

1. UX only, or just additional clarity around the likely state. A major problem now is that the UX implies the user has no posts, despite the possibility that they have many posts but they are simply not visible/loaded. This could be resolved with copy and a more prominent link to the originating server profile. Maybe sufficient; certainly a big improvement.
2. Provide a manual means to directly load in the latest N or last M months of posts. The manual approach here might be a couple orders of magnitude less load on the other server, while still providing clarity and a means to resolve the problem more substantially.
3. Automatically load posts on remote profiles in an efficient/predictable way. This would require far more network design in terms of load on remote servers and more crosstalk if every loaded user profile immediately begins making calls; that's likely not very sustainable, but could be manageable if it say loads the first three posts and waits before loading more; or requires a manual click or scroll before doing so. This one is tough to implement correctly, but would be the seamless experience users would expect from browsing profiles IMO.

In general, we should assume the expectation is that when you visit a user's profile, you see all their posts. Any deviation from that should be treated as unexpected and handled generously and appropriately with messaging and experience accommodations. I expect a balance here is very possible.

**Comment by sneakers-the-rat** [comment]
not that hard to do with a little tidying of the collection handling :) 
https://github.com/mastodon/mastodon/pull/34597

re: perf - i think it's a losing battle to try and regulate on the fetching side. there are things we can/should do to be polite like spacing out requests at some respectful rate, but there's no way to ensure on the fetching side that there is no bad behavior or even accidental coordination (lots of people following and fetching an account after a popular post all at once). It's just got to be handled on the serving side with ratelimits that the fetching side respects, which i believe is how it already works.

**Comment by InkstainTheBat** [comment]
Any updates? I can see the last change to the related PR is from may of this year, could we expect this to be implemented by the end of the year?

**Comment by sneakers-the-rat** [comment]
some statement on that in the last few comments, basically i'm happy to do any additional work to land it whenever, but core dev team is overloaded.
